### ISSUE TITLE
Send campaign to segment of list

### ISSUE THREAD

**Issue by nickwynja** [knadh/listmonk#250]
This feature would allow using queries to send to a segment of a list. This is a pretty standard approach in email service providers. This is somewhat related to #129. The problem I see with creating dynamic "query lists" is that when someone wants to unsubscribe, they might only be unsubscribed from that list and not the "master" list.

To avoid this, I'd like to send a campaign to a segment. For example, I'll add a customer to the master list when they register and include a custom attribute with the subscribe date. Then I could send a campaign to a sub-segment who have registered in, for example, the last 30 days based on a SQL query. The campaign send would still be considered a part of the master list so that an unsubscribe would be unsubscribed from the master list.

Thanks for making listmonk! I've been looking for something like it for a long time.

**Comment by knadh** [comment]
Thanks @nickwynja.

The problem with creating a sub list based on a query is that the results can include subscribers from multiple lists, in which case, it would not possible to have a single parent list.

**Comment by nickwynja** [comment]
I see. I think this could be managed by the segment being a sub-query just like is happening on `/subscribers/lists/<id>` page now. When you query on that page, it appears to only be querying against the list. Here I can search for names or attributes and results are only showing members of the current list.

> Thanks @nickwynja.
> 
> The problem with creating a sub list based on a query is that the results can include subscribers from multiple lists, in which case, it would not possible to have a single parent list.

**Comment by knadh** [comment]
Closing this as there's a longer discussion on the same matter here https://github.com/knadh/listmonk/issues/250.

**Comment by nickwynja** [comment]
@knadh In your comment, you referred to this (self) issue. Did you mean to refer to a different ticket or did you mean to comment and close a different ticket?

**Comment by alerque** [comment]
Please re-open this issue (or link to whatever issue is actually tracking this). It is still a needed feature and the linked discussion issue is ... self referential.

**Comment by knadh** [comment]
@nickwynja @alerque looks like I've lost track of the issue I really meant to reference, and now I'm unable to find it either. Re-opening this.

**Comment by alexbrazier** [comment]
Would it be possible to do something here to filter the list based on a query/segment https://github.com/knadh/listmonk/blob/3386de40c74740b2577e5344e9f6456e9ba50502/queries.sql#L635-L637

```sql
 WHERE subscriber_lists.status != 'unsubscribed' AND
    ## Either grab a query saves in a segments table or get camps.query directly
    ## Not sure if it's possible to convert the string into a real WHERE condition though
    (SELECT query from segments WHERE id = camps.segment) AND
    id > (SELECT last_subscriber_id FROM camps) AND
    id <= (SELECT max_subscriber_id FROM camps)
```

Option would either be to store a new field `query` directly in the campaign - so you would select a `list` and an optional `query` or have a new segments section which would store the query and allow you to view the count of subscribers. Maybe a segments could include the lists so when creating a campaign you would select either a list or a segment.

This would allow us to create separate campaigns for different regions (based on a subscriber attributes) but using a global list.

**Comment by knadh** [comment]
Yep, this is along the lines of the earlier discussions on using sub queries. Campaigns could have a `query` field in the DB (and on the UI). This isn't trivial however. Right now, there's one global query that fetches subscribers in batches for all running campaigns. This will have to change to one query per campaign in addition to the query itself moving to string interpolation to include arbitrary per-campaign query expressions. That also brings in performance considerations.

**Comment by josylad** [comment]
I really hope segmentation will be added to Listmonk soon! 
it is a very important feature for any mailing software.

**Comment by rpmcoding** [comment]
Any eta on list segments?

**Comment by malv007** [comment]
Our use case is the following: Our subscribers receive emails containing events based on their preferences by subscribing to a list. We have a job that runs everyday and populates each subscriber's "attribs" field with a list relevant events, but some days some subscribers may not have any events. Once the "attribs" field has been populated for all users , we create and start a new campaign. The problem we have is that we don't want subscribers to receive an email for a particular day if they don't have any events in their "attribs" property, and right now they are getting an email with no events.

I believe having a "query" property for each campaign will help with something like this. Right?

Do you have any other ideas as to how to implement this using listmonk?

**Comment by satoshinotdead** [comment]
> I really hope segmentation will be added to Listmonk soon! it is a very important feature for any mailing software.

Any update here? Did you make some progress that could be engaged to contribute on code?

It's a must and I suspect that could be a major change on Listmonk scalating steps :)

**Comment by knadh** [comment]
Hi @satoshinotdead. Haven't started working on this yet.

**Comment by satoshinotdead** [comment]
> 

There is a query as workaround to export and import while it's developed? My skills are not good but I like to try and learn.

**Comment by knadh** [comment]
- Create a temporary list
- Use the "Advanced" search to query and segment using SQL expressions. See [docs](https://listmonk.app/docs/querying-and-segmentation/).
- In the results, click on the table checkbox to check one or more subscribers. Use the "Manage lists" bulk option to add the users to the temporary list.
- Send campaign. Once done, delete the temporary list.

**Comment by satoshinotdead** [comment]
Thanks! I have readed before reply because it seems that specific query is not there.

Please, can you give an example about _how can we check open or clicks for `sample_campaign`_?

Big kudos for Listmonk.

**Comment by relikd** [comment]
> * Create a temporary list
> * Use the "Advanced" search [...]

I've seen this suggestion a few times now. What about managing a list of SQL snippets/filter? So you would define a filter once, eg. select all subscribers which do / do-not have a custom attribute "telephone". And, in a campaign, you can optionally select a single filter which is then applied before send. 

I am not sure if this will make implementing easier, but the snippets could be re-used throught the UI.

**Comment by josejuanmontiel** [comment]
Hi @knadh i think the difficult is that POST /api/subscribers/lists send the list of the ids of the subscribers returned by the query... as intro what do you think if we add the compoment of list seleccion and new button "dinamic list" (after test the query)
![imagen](https://github.com/knadh/listmonk/assets/485808/f596afe7-e3ea-47f6-9d8a-f76cb0bff005)
we could add the actual snipped of the query to the table of list (in later iteration we could edit.. in list form) and then ... to update the subscriber dinamicaly... we can launch aun update with the same query to retreive the list ... but now inside NextSubscribers (in pipe.go) update the subscriber_lists table for al this subscribers with the "dinamic list"...

This could be an easy iteration to add "dinamic list" without much overhead... to later consideration:
- Performance (create another scheduler to prepare the data before... we are talking about thousands of subscriber)
- Unsubscribe (with this aproach i think won't be a problem because mantain the same way of use)

For future we can improve:
- The workflow of the segmentation ... other tools create a "temporal tables" for the segmented subscriber, may we can "integrate" dungbeetle ;) in some way
- add some kind of LowCode to deine some workflow bifurcation in segmentation ... mayby using blocky https://codesandbox.io/s/blockly-vue-component-8m460?file=/src/App.vue

But, for the first step i think could be easy. What do you think... open a PR? :)

**Comment by knadh** [comment]
Hi @josejuanmontiel. This indeed warrants a discussion, but it wouldn't be feasible until the next release, which has massive changes (multi-user support). We can pick up this discussion after the v4.0.0 release in a couple of months.

**Comment by alerque** [comment]
This is still a desired feature, loose the bot.

**Comment by skiunke** [comment]
> This is still a desired feature, loose the bot.

So that we're on the same page, this is about creating a campaign and specifying a SQL statement to filter the list the campaigns will be applied to further by their attributes.

**Comment by josejuanmontiel** [comment]
Hi!

> So that we're on the same page, this is about creating a campaign and specifying a SQL statement to filter the list the campaigns will be applied to further by their attributes?

If you ask to me... i don't understand what you said, sorry... my approach, as describe before, will be use the same query that actually return all the user that match (the query) ... to use it exactly before the campaign will launch, in this way it will be "dynamic"... later, could be interesting allow joins with other tables that could be generate in the system by other method for example like dungbeetle project.

**Comment by aniketpanjwani** [comment]
> > This is still a desired feature, loose the bot.
> 
> So that we're on the same page, this is about creating a campaign and specifying a SQL statement to filter the list the campaigns will be applied to further by their attributes?

that's my understanding - send an email (campaign) to part of a list defined by a SQL statement

**Comment by restyler** [comment]
We were about to replace our (insanely resource-heavy) Mautic instance with listmonk. The only crucial feature missing for us is dynamic segments. Our CRM needs to upload subscribers to listmonk and have them automatically assigned to predefined segments. It was disappointing to discover this functionality is absent just as we were ready to make the switch.

Listmonk is an awesome product, and I hope dynamic segment support is implemented soon!

**Comment by restyler** [comment]
Instead of waiting, I took the liberty of submitting a pull request with an important piece of functionality that I believe could bring Listmonk closer to this crucial (and complex) feature: [stored SQL snippets pull request](https://github.com/knadh/listmonk/pull/2546)

**Comment by netapy** [comment]
Any updates on this ?

**Comment by WilliamDEdwards** [comment]
As a new user, I'm really missing this feature. For now, I manually 'segmented' my list by splitting it up into multiple smaller lists, and duplicating the campaign, sending it to a different list every time. :/

Not only do I want to segment to carefully gather recipient feedback without affecting everyone at once, but also because I'm just starting sending newsletters, and need to warm up my domain/IP.

**Comment by argseby** [comment]
What's the status on this issue? This PR https://github.com/knadh/listmonk/pull/2546 made some changes in the direction required for segmentation, but was never merged.

**Comment by WilliamDEdwards** [comment]
> What's the status on this issue? This PR [#2546](https://github.com/knadh/listmonk/pull/2546) made some changes in the direction required for segmentation, but was never merged.

Think you just answered your own question ;)

**Comment by knadh** [comment]
#2546 was too big a PR for me to review and test. It's on my to-do list for the next version.

**Comment by filipesmedeiros** [comment]
I also must say that I wish that we could associate a campaign directly with an SQL query. This does not need #2546, since the person can be asked to write the query, just like they do currently, but for a campaign.

Should I try to help with a PR for that?

Another very viable option that would perhaps be much simpler, would be to add a button to create a list directly from the subscribers page, with the selected subscribers. That would probably be easier to implement, and would at least improve the experience a bit, but it is not the same.

**Comment by fehmi** [comment]
Fingers crossed. Really need that.

**Comment by fehmi** [comment]
I've been using the temporary list workaround suggested by @knadh and ran into some serious issues that I think are worth documenting here.

## My Setup

I have ~17,000 subscribers in a master list, each with JSON attributes for product-based segmentation (e.g. `belong`, `buyed`, `spender`, `edit_image_plan`). I'm using the SQL query UI to filter subscribers and move them to temporary lists for targeted campaigns.

## The Unsubscribe Problem

The temporary list workaround creates a critical unsubscribe flow issue:

**Scenario A — Subscriber is moved to temp list, clicks unsubscribe from an *old* campaign email:**
- The old campaign's `List-Unsubscribe` URL contains the original list's campaign ID
- If the subscriber is no longer in that list (moved to temp), the unsubscribe silently does nothing
- The subscriber receives a success page but is not actually removed from any list
- This is both a deliverability risk and potentially a GDPR/CAN-SPAM compliance issue

**Scenario B — Subscriber clicks unsubscribe from the new campaign (sent via temp list):**
- They are removed from the temp list only
- If you then move them back to the master list after the campaign finishes, their unsubscribed status is lost — they appear as `subscribed` in the master list
- Manual workaround: query `subscriber_lists.status = 'unsubscribed' AND subscriber_lists.list_id = <temp_id>` before moving back, and exclude those users

**Scenario C — One-click unsubscribe (RFC 8058):**
- Gmail and Yahoo require `List-Unsubscribe-Post` for bulk senders
- When sending from a temp list, the unsubscribe URL points to the temp list campaign
- The subscriber is removed from the temp list, not the master list
- If temp list is deleted afterward, the unsubscribe history is gone

## The Reverse Workaround I Tried

To keep unsubscribes on the master list, I explored *reversing* the approach: move the **excluded** segment to a temp list, send the campaign to the master list (which now only contains the target segment), then move the excluded users back after the campaign finishes.

This avoids the unsubscribe URL problem for the *current* campaign, but introduces a race condition: if an excluded subscriber clicks unsubscribe from an *older* campaign email while they are temporarily removed from the master list, the unsubscribe silently fails.

With a slow sending rate (warm-up phase, ~5,000/day), the excluded users can sit in the temp list for **days**, making this window quite large.

## What Would Actually Solve This

Native campaign-level segmentation where:
1. Subscribers remain in the master list at all times
2. A SQL query or segment filter is applied at send time to determine recipients
3. Unsubscribes always reference the master list regardless of which segment received the campaign

This is exactly what tools like Keila handle natively with their segment builder. I'm currently evaluating a migration because the temp list approach introduces compliance risks I can't accept for a production system.

I think the unsubscribe dimension of this issue deserves more attention — it's not just a convenience feature, it has real compliance implications for anyone running a segmented newsletter at scale.

**Comment by filipesmedeiros** [comment]
Bello @fehmi. Thank you so much for your post, awesome stuff!

Regarding your last paragraph, I must say that Keila really is amazing in this aspect: I love the fact that there is only one "list" per project (the Subscribers list) and all segmentation is done with live filters on attributes. I like that approach much more than the list one, which some software use.

Also, the fact that Keila has a visual query builder is very nice. I think it's needed here as well, after the segments are implemented.

Of course, all of this must be followed by: I can help with PRs for whatever is needed!

**Comment by CookedApps** [comment]
We run a newsletter with ~130K subscribers in Listmonk, each with JSON attributes for things like subscription plan, account age etc.

For GDPR we're legally required to scope each campaign to subscribers who consented to that specific purpose. Right now that means pre-populating temp lists via the bulk API before every send, which is stateful and goes stale if attributes change between population and send.

The approach @alexbrazier sketched feels right: an optional `query` field on the campaign, ANDed with the list filter at send time. Unsubscribes stay tied to the list, filtering happens at the campaign level.

Happy to help test this!

**Comment by CookedApps** [comment]
I built a small fork of this: an optional per-campaign `subscriber_query` (the same advanced-query you can already write in the Subscribers view), AND-ed onto the existing list/opt-in filter at send time so it can only narrow recipients within the selected lists. 

Double-opt-in, unsubscribes and blocklist all stay unchanged, and unsubscribe stays bound to the real list. So the List-Unsubscribe problems shouldn't arise.

It reuses what's already there rather than adding surface area: the same `%query%` splice + read-only `EXPLAIN` table-allowlist validation used by the subscriber bulk-query endpoints, gated by the existing `subscribers:sql_query permission`.

Running in production and working well. Happy to open a PR if that's useful.

<img width="3156" height="2494" alt="Image" src="https://github.com/user-attachments/assets/07546f0d-6fae-410e-a10c-5dc1d420fa22" />

**Comment by fehmi** [comment]
> I built a small fork of this: an optional per-campaign `subscriber_query` (the same advanced-query you can already write in the Subscribers view), AND-ed onto the existing list/opt-in filter at send time so it can only narrow recipients within the selected lists.
> 
> Double-opt-in, unsubscribes and blocklist all stay unchanged, and unsubscribe stays bound to the real list. So the List-Unsubscribe problems shouldn't arise.
> 
> It reuses what's already there rather than adding surface area: the same `%query%` splice + read-only `EXPLAIN` table-allowlist validation used by the subscriber bulk-query endpoints, gated by the existing `subscribers:sql_query permission`.
> 
> Running in production and working well. Happy to open a PR if that's useful.
> 
> <img alt="Image" width="2000" height="1580" src="https://private-user-images.githubusercontent.com/10343875/615228414-07546f0d-6fae-410e-a10c-5dc1d420fa22.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODI5MDQ1NzIsIm5iZiI6MTc4MjkwNDI3MiwicGF0aCI6Ii8xMDM0Mzg3NS82MTUyMjg0MTQtMDc1NDZmMGQtNmZhZS00MTBlLWExMGMtNWRjMWQ0MjBmYTIyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA3MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNzAxVDExMTExMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUzOTRkYWE1Y2M5YjU3ZGIyOWUwOTg5MTZhOWJhMmI0YzZjNTMyYzlhY2VkOTkzMTJkZTZiNGI2MDhlOGZkMGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.6_qJ6sFNjZ4oOdHI4j9hUCiDugyTOmchq8iS8aDZTUM">

Looks amazing. Simple is the best. Does it work if multiple lists selected?

**Comment by CookedApps** [comment]
> Looks amazing. Simple is the best. Does it work if multiple lists selected?

Yeah it works with multiple lists. All the selected lists get passed as one `list_id = ANY(...)` set (same as any multi-list send).

**Comment by johnnydecimal** [comment]
As it stands, I'm not sure why I'd bother adding attributes to users. They're far too hard to use.

For any given campaign I need to take `<list> - <users with some attribute> = send to these users`.
